torch_xla/csrc/ops/flash_attention_varlen_position_ids_backward.cpp

```diff
@@ -0,0 +1,434 @@
+                           const torch::lazy::Value& k,
+                           const torch::lazy::Value& v,
+                           const torch::lazy::Value& softmax_lse) {
+  return xla::ShapeUtil::MakeTupleShape(
+      {shape_like(q), shape_like(k), shape_like(v), shape_like(softmax_lse)});
+}
+
+void run_mha_bwd(Flash_bwd_params& params, cudaStream_t stream,
+                 const bool configure) {
+  FP16_SWITCH(!params.is_bf16, [&] {
+    HEADDIM_SWITCH(params.d,
+                   [&] { run_mha_bwd_<elem_type, kHeadDim>(params, stream); });
+  });
+}
+
+// Layout of `buffers` listed above:
+//  buffers[0] = dout
+//  buffers[1] = q
+//  buffers[2] = k
+//  buffers[3] = v
+//  buffers[4] = out
+//  buffers[5] = softmax_lse
+//  buffers[6] = cu_seqlens_q
+//  buffers[7] = cu_seqlens_k
+//  buffers[8] = rng_state
+//  buffers[9] = alibi_slopes
+//  buffers[10] = dq  // this is output
+//  buffers[11] = dk  // this is output
+//  buffers[12] = dv  // this is output
+//  buffers[13] = softmax_d  // this is output
+void custom_call_flash_attention_varlen_position_ids_backward(
+    cudaStream_t stream, void** buffers, const char* opaque,
+    size_t opaque_len) {
+  std::string opaque_str(opaque, opaque_len);
+  TF_VLOG(3)
+      << "custom_call_flash_attention_varlen_position_ids_backward opaque str: "
+      << opaque_str;
+  FlashAttentionBackwardParams params;
+  params.FromString(std::move(opaque_str));
+  int buf_offset = params.enable_alibi_slopes;
+  auto scalar_type = params.is_bf16 ? torch::kBFloat16 : torch::kFloat16;
+
+  cudaStream_t torch_stream = at::cuda::getCurrentCUDAStream().stream();
+  cudaEvent_t torch_wait_xla_event;
+  cudaEventCreateWithFlags(&torch_wait_xla_event, cudaEventDisableTiming);
+  cudaEvent_t xla_wait_torch_event;
+  cudaEventCreateWithFlags(&xla_wait_torch_event, cudaEventDisableTiming);
+  cudaEventRecord(torch_wait_xla_event, stream);
+  cudaStreamWaitEvent(torch_stream, torch_wait_xla_event);
+
+  auto cuda_stream = at::cuda::getCurrentCUDAStream();
+  at::cuda::CUDAStreamGuard guard(cuda_stream);
+
+  auto opts = torch::TensorOptions().dtype(scalar_type).device(torch::kCUDA);
+
+  // Inputs
+  at::Tensor do_ = torch::from_blob(
+      buffers[0], {params.b * params.seqlen_q, params.h, params.d}, opts);
+  at::Tensor q = torch::from_blob(
+      buffers[1], {params.b * params.seqlen_q, params.h, params.d}, opts);
+  at::Tensor k = torch::from_blob(
+      buffers[2], {params.b * params.seqlen_k, params.h_k, params.d}, opts);
+  at::Tensor v = torch::from_blob(
+      buffers[3], {params.b * params.seqlen_k, params.h_k, params.d}, opts);
+  at::Tensor o = torch::from_blob(
+      buffers[4], {params.b * params.seqlen_q, params.h, params.d}, opts);
+  at::Tensor softmax_lse =
+      torch::from_blob(buffers[5], {params.b, params.h, params.seqlen_q},
+                       opts.dtype(torch::kFloat));
+  at::Tensor cu_seqlens_q = torch::from_blob(buffers[6], {params.seqlen_q + 1},
```

Comment:
Init with size `params.b * params.seqlen + 1` incase batch_size >1 in the future?

Reply:
done